Source/Core/Core/IOS/USB/Emulated/Skylander.cpp

```diff
@@ -0,0 +1,751 @@
+
+        // The 6th byte {3e} is a counter and increments by one. It will roll over when reaching
+        // {ff}.
+
+        // The purpose of the (00\|01) byte at the 7th position appear to indicate if the portal has
+        // been activated: {01} when active and {00} when deactivated.
+      case 'S':
+      {
+        q_data = {buf[0]};
+        cmd->expected_count = 9;
+        break;
+      }
+      case 'V':
+      {
+        q_data = {buf[0], buf[1], buf[2], buf[3]};
+        cmd->expected_count = 12;
+        break;
+      }
+        // Write
+        // Command	{ 'W', 10, 00, 01, 02, 03, 04, 05, 06, 07, 08, 09, 0a, 0b, 0c, 0d, 0e, 0f, 00,
+        // 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00 }
+        // Response	{ 'W', 00, 00, 00, 00, 00,
+        // 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00,
+        // 00, 00, 00, 00 }
+        // In the command the 3rd byte indicates which Skylander to query data from. Index starts at
+        // 0x10 for the 1st Skylander (as reported in the Status command.) The 16th Skylander
+        // indexed would be 0x20.
+
+        // 4th byte is the block to write to.
+
+        // Bytes 5 - 20 ({ 01, 02, 03, 04, 05, 06, 07, 08, 09, 0a, 0b, 0c, 0d, 0e, 0f }) are the
+        // data to write.
+
+        // The response does not appear to return the id of the Skylander being written, the 2nd
+        // byte is 0x00; however, the 3rd byte echos the block that was written (0x00 in example
+        // above.)
+
+      case 'W':
+      {
+        const u8 sky_num = buf[1] & 0xF;
+        const u8 block = buf[2];
+        g_skyportal.WriteBlock(sky_num, block, &buf[3], q_result.data());
+        q_queries.push(q_result);
+        q_data = {buf[0],  buf[1],  buf[2],  buf[3],  buf[4],  buf[5],  buf[6],
+                  buf[7],  buf[8],  buf[9],  buf[10], buf[11], buf[12], buf[13],
+                  buf[14], buf[15], buf[16], buf[17], buf[18]};
+        cmd->expected_count = 19;
+        break;
+      }
+      default:
+        ERROR_LOG_FMT(IOS_USB, "Unhandled Skylander Portal Query: {}", buf[0]);
+        break;
+      }
+      break;
+    case 0x0A:
+      cmd->expected_count = 8;
+      cmd->OnTransferComplete(8);
+      return 0;
+    case 0x0B:
+      cmd->expected_count = 8;
+      cmd->OnTransferComplete(8);
+      return 0;
+    default:
+      ERROR_LOG_FMT(IOS_USB, "Unhandled Request {}", cmd->request);
+      break;
+    }
+    break;
+
+  default:
+    break;
+  }
+  cmd->expected_time = Common::Timer::NowUs() + 100;
+  GetTransferThread().AddTransfer(std::move(cmd), q_data);
+  return 0;
+}
+int SkylanderUsb::SubmitTransfer(std::unique_ptr<BulkMessage> cmd)
+{
+  DEBUG_LOG_FMT(IOS_USB, "[{:04x}:{:04x} {}] Bulk: length={} endpoint={}", m_vid, m_pid,
+                m_active_interface, cmd->length, cmd->endpoint);
+  return 0;
+}
+int SkylanderUsb::SubmitTransfer(std::unique_ptr<IntrMessage> cmd)
```

Comment:
Sure, that particular SubmitTransfer method handles Interrupt transfers. Interrupt transfers is how the Portal communicates information back to the console, where the other SubmitTransfer method with the long Switch/Case block is the Control Transfers. The documentation for this is online, I'll link [this](https://www-user.tu-chemnitz.de/~heha/hsn/chm/usb.chm/usb4.htm#:~:text=Bulk%20Transfers-,Control%20Transfers,and%20use%20best%20effort%20delivery.) article to explain better, but basically the console sends out Control Transfers to the Skylander Portal, and the response to those is just echoed back to acknowledge the transfer was received. In the Interrupt Message version of SubmitTransfer, any data that was requested via the Control Transfer is returned by the response to these commands. For example, when the portal is activated, the ControlMessage method echoes back the request, but then the next InterruptMessage that is received then responds with the requested data. You can see this behaviour in line 420:

`if (!q_queries.empty())
  {
    q_result = q_queries.front();
    q_queries.pop();
    // This needs to happen after ~22 milliseconds
    cmd->expected_time = Common::Timer::NowUs() + 22000;
  }
  else
  {
    q_result = g_skyportal.GetStatus();
    cmd->expected_time = Common::Timer::NowUs() + 2000;
  }`

If there are any requests that need a response, they are taken from the top of the queue of results and responded to after ~22 milliseconds, otherwise the response is just a status update from the portal (after about 2ms, behaviour shown from the real portal)

TLDR; Control Transfers are just echoed back to the system, but adds an expected response for upcoming Interrupt Transfers, and Interrupt Transfers respond with the data requested by the Control Transfer

Reply:
Unless you were referring to the BulkMessage method, in which case that is only there as you need to implement that Method if you extend the Device class

Reply:
Okay. I guessed so. Just found the code a bit confusing. Thanks